core/src/main/scala/org/apache/spark/serializer/SerializationDebugger.scala

```diff
@@ -169,18 +208,53 @@ private[spark] object SerializationDebugger extends Logging {
             }
             j += 1
           }
-
         }
         i += 1
       }
       return List.empty
     }
+
+    /**
+     * Visit a serializable object which has the writeObject() defined.
+     * Since writeObject() can choose add arbitrary objects at the time of serialization,
+     * the only way to capture all the objects it will serialize is by using a
+     * dummy ObjectOutputStream that collects all the relevant fields for further testing.
```

Comment:
maybe it's worth mentioning that this is the similar to how externalizable objects are visited?